### PR TITLE
Boost serialization implementation

### PR DESCRIPTION
**Pull request by ron0studios**
A simple boost.serialization custom format for rfl. 

Currently the basics are supported, but I feel adding patches will be outside the scope of this very first PR. 

TODO:
- docs
- better/more rigorous tests (the current ones are just for validation). A test that showcases the Archive system of boost serialisation would be nice (&/<</>> overloads)
- other boost formats (text_oarchive, xml_oarchive)
- nested structs, rfl::Rename, unique/shared ptrs (box/ref), and a couple other rfl features.
- better tests will probably reveal bugs :)

### CODE REVIEW

**Comment by ron0studios** [comment]
undid the gemini code assist suggestion ([57031b0](https://github.com/getml/reflect-cpp/commit/57031b083d4b6a2f745424e1accc4d4868e4568e))

File: include/rfl/boost_serialization/Reader.hpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+  template <class ObjectReader>
+  std::optional<Error> read_object(const ObjectReader& _object_reader,
+                                   const InputObjectType& _obj) const noexcept {
+    for (size_t i = 0; i < _obj.size; ++i) {
+      auto var = InputVarType{_obj.ar};
+      _object_reader.read(static_cast<int>(i), var);
+    }
+    return std::nullopt;
+  }
```
[the remark this answers is not in the document]
Comment (ron0studios) [reply include/rfl/boost_serialization/Reader.hpp:162]: viewreader::read returns void, similar to how capnproto and avro handle it. Perhaps this could be addressed in a later pr? I think viewreader has an internal error collection scheme too

**Comment by liuzicheng1987** [comment]
@ron0studios let me know when this is ready for review

**Comment by liuzicheng1987** [comment]
@ron0studios I just had a quick look. I think you're on the right track. I look forward to having this.

**Comment by ron0studios** [comment]
@liuzicheng1987 Ready for review - everything but docs is written through. The remaining work was a matter of writing tests to see if some edge cases already worked out the box or not :)

**Comment by liuzicheng1987** [comment]
@ron0studios , first of all, thanks for the contribution. This looks great and it inspired me to support https://github.com/USCiLab/cereal next. 

I have a couple of minor remarks:

1. As you have already observed, this still needs documentation. It needs to be mentioned in the README and it needs its own section in the general documentation. Feel free to write that using a coding agent of your choice (if you don't have one, I can do that).

2. You need to update the Github Actions files to actually run your tests. I was able to run them locally and they pass, of course, but they need to run in an automated way.

3. I think you could outsource some of the functionality (such as writer) into source files. Compile time is a major issue in a project like this and anything we can do to reduce it, we should do.

4. It would also be cool to include this in the benchmarks. I am actually very curious how this performs in comparison to other serialization formats.

5. I think we could add a few more tests and it wouldn't be hard to do...you could simply copy the tests for BSON, say, and they just search + replace bson with boost_serialization.

As a final remark, I am a bit concerned about security here, as there appears to be next to no validation...I wonder to what extent it would be possible to add a malicious file to get the code to do things it's not supposed. But this seems to be a general problem with Boost.Serialization and it's certainly not our job to fix it.

If you want to me to take over any of the points 1-5, I'd be happy to do so.

**Comment by ron0studios** [comment]
@liuzicheng1987 thanks for reading through!

I'm happy to go through 1-5 :)

I have never tackled the security side of things before, so I'm not sure how helpful I can be on that front unfortunately.

I'll keep you posted!

**Comment by ron0studios** [comment]
Screenshot 2026-04-03 at 10 41 36

some benchmarking. I think I could make this faster by reducing the copies on the read path.

**Comment by liuzicheng1987** [comment]
@ron0studios, I have since added the cereal support, so please merge main once and resolve the merge conflicts in the CMakeLists (which should be very easy).

**Comment by liuzicheng1987** [comment]
@ron0studios other than that, looks good to me. If the tests run smoothly, I see no reason not to merge this, unless you want to add your optimization first.

**Comment by liuzicheng1987** [comment]
@ron0studios, the tests look good. The two failures are obviously a server issue, nothing to do with your code. So if you could merge main, that would be great. I would then merge the PR, unless you want to do the optimization first.

**Comment by ron0studios** [comment]
rebased against origin/main just earlier @liuzicheng1987. Happy to get this merge through :)

**Comment by liuzicheng1987** [comment]
@ron0studios merged. Thanks for your contribution!
